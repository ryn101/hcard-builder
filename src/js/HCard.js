/**
 * Container for HCard specific information
 */
class HCard {
    constructor(
        givenName,
        surname,
        email,
        phone,
        houseNumber,
        street,
        suburb,
        state,
        postcode,
        country,
    ) {
        this.personalDetails = {
            givenName,
            surname,
            email,
            phone,
        };

        this.address = {
            houseNumber,
            street,
            suburb,
            state,
            postcode,
            country,
        };

        Object.defineProperty(this, 'fullName', {
            enumerable: false,
            get: () => [this.personalDetails.givenName, this.personalDetails.surname].filter(Boolean).join(' '),
        });
    }

    /**
     * Returns the full name being a combination of given and surnames
     *
     * @returns {string}
     */
    fullName = () => {
        const { givenName, surname } = this.personalDetails;
        return [givenName, surname].filter(Boolean).join(' ');
    }
}

export default HCard;
