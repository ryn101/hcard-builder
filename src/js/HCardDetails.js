/**
 * Container for HCard detail information
 *
 * @property givenName - The first/given name
 * @property surname - The surname
 * @property email - The email
 * @property phone - The mobile/home/work phone number
 * @property houseNameOrNumber - The house/unit number/name
 * @property street - The street address
 * @property suburb - The address suburb
 * @property state - The address state
 * @property postcode - The address postcode
 * @property country - The address country
 */
class HCardDetails {
    constructor(
        givenName,
        surname,
        email,
        phone,
        houseNameOrNumber,
        street,
        suburb,
        state,
        postcode,
        country,
    ) {
        this.givenName = givenName;
        this.surname = surname;
        this.email = email;
        this.phone = phone;
        this.houseNameOrNumber = houseNameOrNumber;
        this.street = street;
        this.suburb = suburb;
        this.state = state;
        this.postcode = postcode;
        this.country = country;
    }
}

export default HCardDetails;
