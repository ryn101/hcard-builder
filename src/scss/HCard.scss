@import 'colours';
@import 'mixins';

$vcard-padding: 25px;
$photo-wrapper-right: 10px;
$photo-wrapper-width: 83px;

.vcard {
  width: 430px;
  background-color: white;
  border-bottom: 2px solid #9A9A9A;
  height: 306px;
  position: relative;

  .type {
    display: none;
  }

  .banner {
    position: absolute;
    top: -26px;
    right: 5px;
    color: #95A5A6;
  }

  .header {
    height: 100px;
    background-color: $sophisticated-navy;
    padding: 54px $vcard-padding 0 $vcard-padding;

    h1 {
      font-size: 18pt;
      margin: 0;
      color: #fff;
      font-family: Merriweather, Arial, serif;
      width: calc(100% - #{($photo-wrapper-right * 2) + $photo-wrapper-width - $vcard-padding});
      @include ellipsis();
    }
  }

  .photo-wrapper {
    border: 1px solid #9A9A9A;
    border-right-width: 2px;
    border-bottom-width: 2px;
    position: absolute;
    right: $photo-wrapper-right;
    top: 10px;
    height: 104px;
    width: $photo-wrapper-width;
    overflow: hidden;
    text-align: center;
    background-color: #e3e3e3;

    > img {
      height: 100%;
      width: 100%;
      object-fit: cover;
    }
  }

  .details {
    display: block;
    padding: $vcard-padding;
    margin: 0;
    font-style: normal;

    li {
      border-bottom: 1px $on-the-line-grey solid;
      padding-bottom: 1px;
      justify-content: left;

      &:not(:first-child) {
        margin-top: 12px;
      }
    }

    .detail {
      width: 100%;
      overflow: hidden;

      div > span {
        display: inline-block;
      }

      $detail-label-width: 76px;

      .label {
        width: $detail-label-width;

        color: $sophisticated-navy;
        text-transform: uppercase;
        font-size: 7.5pt;
      }

      .value {
        @include ellipsis();
        font-family: Merriweather, Arial, serif;
        width: calc(100% - #{$detail-label-width});
        vertical-align: text-bottom;
        margin-bottom: -5px;
        padding-bottom: 5px;
        min-height: 23px;
      }
    }
  }
}
