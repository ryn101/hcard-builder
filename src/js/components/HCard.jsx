import React from 'react';
import PropTypes from 'prop-types';
import HCardDetails from '../HCardDetails';

import '../../scss/HCard.scss';
import profilePicture from '../../images/profile-default.png';
import HCardProperty from './HCardProperty';
import HCardDetail from './HCardDetail';
import AddressState from '../AddressState';

function HCard(props) {
    const { hCard, avatarFileUrl } = props;
    const streetAddress = [hCard.houseNameOrNumber, hCard.street].filter(Boolean).join(' ');
    const fullName = [hCard.givenName, hCard.surname].filter(Boolean).join(' ');
    const state = new AddressState(hCard.state);

    return (
        <section className="vcard">
            <header className="vcard__header">
                <h2 className="vcard__heading fn" title={fullName}>{fullName}</h2>
                <div className="vcard__photo-wrapper">
                    <img alt="avatar" className="photo" src={avatarFileUrl} />
                </div>
            </header>
            <span className="vcard__banner">HCARD PREVIEW</span>
            <address className="adr">
                <ul className="vcard__details">
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Email" className="flex-row__column">
                            <HCardProperty name="email" value={hCard.email} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Phone" className="flex-row__column">
                            <HCardProperty name="tel">
                                <HCardProperty name="type" value="home" />
                                { hCard.phone }
                            </HCardProperty>
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Address" className="flex-row__column">
                            <HCardProperty name="street-address" value={streetAddress} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail className="flex-row__column">
                            <HCardProperty name="locality" value={hCard.suburb} />
                            <span>{hCard.suburb && hCard.state ? ', ' : ''}</span>
                            <HCardProperty name="region" value={state.name} abbreviation={state.abbreviation} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-2">
                        <HCardDetail label="Post Code" className="flex-row__column">
                            <HCardProperty name="postal-code" value={hCard.postcode} />
                        </HCardDetail>
                        <HCardDetail label="Country" className="flex-row__column">
                            <HCardProperty name="country" value={hCard.country} />
                        </HCardDetail>
                    </li>
                </ul>
            </address>
        </section>
    );
}

HCard.propTypes = {
    hCard: PropTypes.instanceOf(HCardDetails).isRequired,
    avatarFileUrl: PropTypes.string,
};

HCard.defaultProps = {
    avatarFileUrl: profilePicture,
};

export default HCard;
