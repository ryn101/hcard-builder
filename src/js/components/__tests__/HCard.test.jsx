import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HCardDetails from '../../HCardDetails';
import HCard from '../HCard';

configure({ adapter: new Adapter() });

it('renders the hCard without crashing', () => {
    const hCard = new HCardDetails();
    const div = document.createElement('div');
    ReactDOM.render(<HCard hCard={hCard} />, div);
    ReactDOM.unmountComponentAtNode(div);
});

describe('once rendered', () => {
    let hCardNode;
    let hCard;

    beforeEach(() => {
        hCard = new HCardDetails();
        hCardNode = shallow(<HCard hCard={hCard} />);
    });

    it('renders with state correctly', () => {
        expect(hCardNode).toMatchSnapshot();
    });

    it('renders the full name correctly', () => {
        const givenName = 'Ryan';
        const surname = 'O\'Connor';

        hCard.givenName = givenName;
        hCard.surname = surname;
        hCardNode.setProps({ hCard });

        const fullNameNode = hCardNode.find('.fn');
        expect(fullNameNode.text()).toEqual(`${givenName} ${surname}`);
    });

    it('set the avatar file url', () => {
        const avatarFileUrl = '/images/test.png';
        hCardNode.setProps({ avatarFileUrl });
        hCardNode.update();

        const photoNode = hCardNode.find('.photo');
        expect(photoNode.prop('src')).toEqual(avatarFileUrl);
    });
});
