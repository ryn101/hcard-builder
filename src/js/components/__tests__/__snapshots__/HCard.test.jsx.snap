// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once rendered renders with state correctly 1`] = `
<HCard
  avatarFileUrl="test-file-stub"
  hCard={
    HCardDetails {
      "country": undefined,
      "email": undefined,
      "givenName": undefined,
      "houseNameOrNumber": undefined,
      "phone": undefined,
      "postcode": undefined,
      "state": undefined,
      "street": undefined,
      "suburb": undefined,
      "surname": undefined,
    }
  }
>
  <section
    className="vcard"
  >
    <header
      className="vcard__header"
    >
      <HCardProperty
        as="h2"
        className="vcard__heading"
        name="fn"
        value=""
      >
        <h2
          className="fn vcard__heading"
          title=""
        />
      </HCardProperty>
      <figure
        className="vcard__photo-wrapper"
      >
        <img
          alt="avatar"
          className="photo"
          src="test-file-stub"
        />
      </figure>
    </header>
    <span
      className="vcard__banner"
    >
      HCARD PREVIEW
    </span>
    <HCardProperty
      as="dl"
      className="vcard__details"
      name="adr"
      value=""
    >
      <dl
        className="adr vcard__details"
        title=""
      >
        <HCardDetail
          columns={1}
          label="Email"
        >
          <div
            className="detail detail--columns-1"
          >
            <dt
              className="detail__label"
            >
              Email
            </dt>
            <dd
              className="detail__value"
            >
              <HCardProperty
                as="span"
                className=""
                name="email"
                value=""
              >
                <span
                  className="email"
                  title=""
                />
              </HCardProperty>
            </dd>
          </div>
        </HCardDetail>
        <HCardDetail
          columns={1}
          label="Phone"
        >
          <div
            className="detail detail--columns-1"
          >
            <dt
              className="detail__label"
            >
              Phone
            </dt>
            <dd
              className="detail__value"
            >
              <HCardProperty
                as="span"
                className=""
                name="tel"
                value=""
              >
                <span
                  className="tel"
                  title=""
                >
                  <HCardProperty
                    as="span"
                    className=""
                    name="type"
                    value="home"
                  >
                    <span
                      className="type"
                      title="home"
                    >
                      home
                    </span>
                  </HCardProperty>
                </span>
              </HCardProperty>
            </dd>
          </div>
        </HCardDetail>
        <HCardDetail
          columns={1}
          label="Address"
        >
          <div
            className="detail detail--columns-1"
          >
            <dt
              className="detail__label"
            >
              Address
            </dt>
            <dd
              className="detail__value"
            >
              <HCardProperty
                as="span"
                className=""
                name="street-address"
                value=""
              >
                <span
                  className="street-address"
                  title=""
                />
              </HCardProperty>
            </dd>
          </div>
        </HCardDetail>
        <HCardDetail
          columns={1}
          label={null}
        >
          <div
            className="detail detail--columns-1"
          >
            <dt
              className="detail__label"
            />
            <dd
              className="detail__value"
            >
              <HCardProperty
                as="span"
                className=""
                name="locality"
                value=""
              >
                <span
                  className="locality"
                  title=""
                />
              </HCardProperty>
              <span />
              <HCardProperty
                abbreviation={null}
                as="span"
                className=""
                name="region"
                value=""
              >
                <span
                  className="region"
                  title=""
                />
              </HCardProperty>
            </dd>
          </div>
        </HCardDetail>
        <HCardDetail
          columns={2}
          label="Post Code"
        >
          <div
            className="detail detail--columns-2"
          >
            <dt
              className="detail__label"
            >
              Post Code
            </dt>
            <dd
              className="detail__value"
            >
              <HCardProperty
                as="span"
                className=""
                name="postal-code"
                value=""
              >
                <span
                  className="postal-code"
                  title=""
                />
              </HCardProperty>
            </dd>
          </div>
        </HCardDetail>
        <HCardDetail
          columns={2}
          label="Country"
        >
          <div
            className="detail detail--columns-2"
          >
            <dt
              className="detail__label"
            >
              Country
            </dt>
            <dd
              className="detail__value"
            >
              <HCardProperty
                as="span"
                className=""
                name="country"
                value=""
              >
                <span
                  className="country"
                  title=""
                />
              </HCardProperty>
            </dd>
          </div>
        </HCardDetail>
      </dl>
    </HCardProperty>
  </section>
</HCard>
`;
